### Definição algoritmo UPMGA

In [30]:
import pandas as pd
matrix = pd.read_csv('matriz_distancias.csv', sep = '\t', index_col= 'index')

def encontraMenorDistancia(matrix):
  min = matrix.iloc[0,-1]
  for i in matrix.iterrows():
    for j in matrix:
      if (i[1][j] >= min and i[1][j] < 0): #Essa parte é diferente da questão 1
        min = i[1][j]
        gen1 = i[0]
        gen2 = j
  return min, (gen1, gen2)

def substituiPar(matrix, par):
    print(matrix)
    matrix[str(par).replace("'",'')] = matrix.apply(lambda x: (x[par[0]] + x[par[1]])/2)
    matrix.drop(list(par), axis =1, inplace=True)
    matrix.drop(list(par), inplace = True)
    vec = list(matrix[str(par).replace("'",'')])
    vec.append(0)
    matrix.loc[str(par).replace("'",'')] = vec
    print(min)
    

def upmga(matrix):
  while len(matrix.columns) > 1:
    min, par = encontraMenorDistancia(matrix)
    substituiPar(matrix, par)
  return (matrix.columns[0])
  

### Questão 2 - A

|Sequencia|Organismo|
|---|:---:|
|seq_1|Bacillus anthracis|
|seq_2|Clostridium tetani|
|seq_3|Escherichia coli|
|seq_4|Mycobacterium tuberculosis|
|seq_5|Pseudomonas aerugin|
|seq_6|Salmonella enteric|
|seq_7|Staphylococcus aureus|
|seq_8|Streptococcus pneumoni|
|seq_9|Treponema pallidum|
|seq_10|Vibrio cholerae|

### Questão 2 - B

In [31]:
import sys

gap = -2
mismatch = -1
match = 1

def calculaScore(seq1, seq2):
  if len(seq2) > len(seq1):
      seq1, seq2 = seq2, seq1
  i = 1
  j = 0
  seq1 = ' '+ seq1
  seq2 = ' ' + seq2
  linha_anterior = list(range(0,(len(seq1))*gap, gap))
  while i < len(seq1):
    atual = [(i)*gap]
    for j in range(1,len(seq2)):
      if seq1[i] == seq2[j]:
        atual.append(max(linha_anterior[j -1]+match, linha_anterior[j] + gap, atual[-1] + gap))
      else:
        atual.append(max(linha_anterior[j -1]+mismatch, linha_anterior[j] + gap, atual[-1] + gap))
    linha_anterior = atual
    i += 1
  return atual[-1]



In [32]:

seqs= []
for i in range(1,11):
    seqs.append(open(f'seq_{i}.fna', 'r').read().replace('\n', '')[1:])

seqs[0] = seqs[0][9482:9732]
seqs[1] = seqs[1][177903:178153]
seqs[2] = seqs[2][3519919:3520169]
seqs[3] = seqs[3][1490249:1490499]
seqs[4] = seqs[4][731123:731373]
seqs[5] = seqs[5][682942:683192]
seqs[6] = seqs[6][454284:454434]
seqs[7] = seqs[7][15079:15329]
seqs[8] = seqs[8][234103:234353]
seqs[9] = seqs[9][253:503]


resultados = []
for i, seq_a in enumerate(seqs):
    resultados.append([])
    for j, seq_b in enumerate(seqs):
        if i < j:
            result = calculaScore(seq_a, seq_b)
            i_ = i+1
            j_ = j+1
            resultados[i].append(result)
            print(f'seq_{i_} vs seq_{j_}: {result}')

for i in range(10):
    resultados[i] = [0] + resultados[i]
    for j in range(i):
        resultados[i] = [0] + resultados[i]
for i in range(10):
    for j in range(i):
        resultados[i][j] = resultados[j][i]

resultados
    

seq_1 vs seq_2: -49
seq_1 vs seq_3: -30
seq_1 vs seq_4: -40
seq_1 vs seq_5: -33
seq_1 vs seq_6: -51
seq_1 vs seq_7: -124
seq_1 vs seq_8: -33
seq_1 vs seq_9: -39
seq_1 vs seq_10: -23
seq_2 vs seq_3: -58
seq_2 vs seq_4: -43
seq_2 vs seq_5: -38
seq_2 vs seq_6: -49
seq_2 vs seq_7: -124
seq_2 vs seq_8: -29
seq_2 vs seq_9: -39
seq_2 vs seq_10: -45
seq_3 vs seq_4: -29
seq_3 vs seq_5: -25
seq_3 vs seq_6: -36
seq_3 vs seq_7: -119
seq_3 vs seq_8: -37
seq_3 vs seq_9: -31
seq_3 vs seq_10: -44
seq_4 vs seq_5: -21
seq_4 vs seq_6: -21
seq_4 vs seq_7: -134
seq_4 vs seq_8: -29
seq_4 vs seq_9: -37
seq_4 vs seq_10: -42
seq_5 vs seq_6: -12
seq_5 vs seq_7: -128
seq_5 vs seq_8: -29
seq_5 vs seq_9: -45
seq_5 vs seq_10: -28
seq_6 vs seq_7: -144
seq_6 vs seq_8: -29
seq_6 vs seq_9: -47
seq_6 vs seq_10: -52
seq_7 vs seq_8: -121
seq_7 vs seq_9: -124
seq_7 vs seq_10: -121
seq_8 vs seq_9: -39
seq_8 vs seq_10: -38
seq_9 vs seq_10: -37


[[0, -49, -30, -40, -33, -51, -124, -33, -39, -23],
 [-49, 0, -58, -43, -38, -49, -124, -29, -39, -45],
 [-30, -58, 0, -29, -25, -36, -119, -37, -31, -44],
 [-40, -43, -29, 0, -21, -21, -134, -29, -37, -42],
 [-33, -38, -25, -21, 0, -12, -128, -29, -45, -28],
 [-51, -49, -36, -21, -12, 0, -144, -29, -47, -52],
 [-124, -124, -119, -134, -128, -144, 0, -121, -124, -121],
 [-33, -29, -37, -29, -29, -29, -121, 0, -39, -38],
 [-39, -39, -31, -37, -45, -47, -124, -39, 0, -37],
 [-23, -45, -44, -42, -28, -52, -121, -38, -37, 0]]

|Sequencia A|Sequencia B|Score|
|:---:|:---:|:---:|
|seq_1 | seq_2| -49|
|seq_1 | seq_3| -30|
|seq_1 | seq_4| -40|
|seq_1 | seq_5| -33|
|seq_1 | seq_6| -51|
|seq_1 | seq_7| -124|
|seq_1 | seq_8| -33|
|seq_1 | seq_9| -39|
|seq_1 | seq_10| -23|
|seq_2 | seq_3| -58|
|seq_2 | seq_4| -43|
|seq_2 | seq_5| -38|
|seq_2 | seq_6| -49|
|seq_2 | seq_7| -124|
|seq_2 | seq_8| -29|
|seq_2 | seq_9| -39|
|seq_2 | seq_10| -45|
|seq_3 | seq_4| -29|
|seq_3 | seq_5| -25|
|seq_3 | seq_6| -36|
|seq_3 | seq_7| -119|
|seq_3 | seq_8| -37|
|seq_3 | seq_9| -31|
|seq_3 | seq_10| -44|
|seq_4 | seq_5| -21|
|seq_4 | seq_6| -21|
|seq_4 | seq_7| -134|
|seq_4 | seq_8| -29|
|seq_4 | seq_9| -37|
|seq_4 | seq_10| -42|
|seq_5 | seq_6| -12|
|seq_5 | seq_7| -128|
|seq_5 | seq_8| -29|
|seq_5 | seq_9| -45|
|seq_5 | seq_10| -28|
|seq_6 | seq_7| -144|
|seq_6 | seq_8| -29|
|seq_6 | seq_9| -47|
|seq_6 | seq_10| -52|
|seq_7 | seq_8| -121|
|seq_7 | seq_9| -124|
|seq_7 | seq_10| -121|
|seq_8 | seq_9| -39|
|seq_8 | seq_10| -38|
|seq_9 | seq_10| -37|

### Questão 2 - C

In [27]:
pd.read_csv('matriz_distancias_questao_2.csv', sep=';', index_col= 'index')

seq1  seq2  seq3  seq4  seq5  seq6  seq7  seq8  seq9  seq10
index                                                             
seq1      0   -49   -30   -40   -33   -51  -124   -33   -39    -23
seq2    -49     0   -58   -43   -38   -49  -124   -29   -39    -45
seq3    -30   -58     0   -29   -25   -36  -119   -37   -31    -44
seq4    -40   -43   -29     0   -21   -21  -134   -29   -37    -42
seq5    -33   -38   -25   -21     0   -12  -128   -29   -45    -28
seq6    -51   -49   -36   -21   -12     0  -144   -29   -47    -52
seq7   -124  -124  -119  -134  -128  -144     0  -121  -124   -121
seq8    -33   -29   -37   -29   -29   -29  -121     0   -39    -38
seq9    -39   -39   -31   -37   -45   -47  -124   -39     0    -37
seq10   -23   -45   -44   -42   -28   -52  -121   -38   -37      0

### Questão 2 - D

In [28]:
matrix = pd.read_csv('matriz_distancias_questao_2.csv', sep=';', index_col= 'index')
upmga(matrix)

       seq1  seq2  seq3  seq4  seq5  seq6  seq7  seq8  seq9  seq10
index                                                             
seq1      0   -49   -30   -40   -33   -51  -124   -33   -39    -23
seq2    -49     0   -58   -43   -38   -49  -124   -29   -39    -45
seq3    -30   -58     0   -29   -25   -36  -119   -37   -31    -44
seq4    -40   -43   -29     0   -21   -21  -134   -29   -37    -42
seq5    -33   -38   -25   -21     0   -12  -128   -29   -45    -28
seq6    -51   -49   -36   -21   -12     0  -144   -29   -47    -52
seq7   -124  -124  -119  -134  -128  -144     0  -121  -124   -121
seq8    -33   -29   -37   -29   -29   -29  -121     0   -39    -38
seq9    -39   -39   -31   -37   -45   -47  -124   -39     0    -37
seq10   -23   -45   -44   -42   -28   -52  -121   -38   -37      0
<built-in function min>
               seq1   seq2   seq3   seq4   seq7   seq8   seq9  seq10  \
index                                                                  
seq1            0.0  -49.0  

'((((seq9, seq3), (seq10, seq1)), ((((seq6, seq5), seq4), seq8), seq2)), seq7)'